Lesson: React Ecosystem - React Router (#25040)
TheOdinProject/curriculum: react/the_react_ecosystem/react_router.md

- How do you add protected routes?

### Client-Side Routing

Client-side routing is the type of routing where Javascript takes over the duty of handling the routes in an application. Client-side routing helps in building single-page applications (SPAs) without refreshing as the user navigates. For example, when a user clicks a navbar element, the URL changes and the view of the page is modified accordingly, within the client.

Say you are cooking some chicken. If you want to cook it well and nice, you will have to:

1. Put the chicken in the oven and set it to cook with appropriate time and heating
2. Wait till the dish gives out that satisfying smell
3. Start munching!

This is common to all websites, you set the oven up for what you want (visit any URL, like https://theodinproject.com/), wait for the oven to be done with the cooking (the loading screen), and tada, enjoy your delicious food (your page is ready for use). But what if you forgot to add some spices before you cooked it up? You have to repeat this flow again:

1. Add the spices to the chicken
2. Put it back into the oven and set it up to be reheated
3. Wait for it to be nice and warm
4. Now you can eat it!

> Here is where we reiterate, **the chicken needs to be reheated**. In a general multi-page application (MPAs), the browser reloads every time you click on a link to navigate. With client-side routing, **you never leave the page you are on** - you bring the microwave to the table to ensure that you don't run into the "missing spices" issues. The data is rendered with Javascript, instead of letting the browser handle it.

Author: Rewritten to line 32. 
"The link requests are intercepted by the Javascript that you write, instead of letting them go directly to the server."

### A Reactive Solution

While client-side routing allows for nicer, app-like interactions (since you are controlling the routes, you can make fancy CSS animations across route changes), a lot of caveats can be missed. Browsers reloads notify screen-readers of new content to read, so you will need to notify screen-readers of route updates manually. However, with the help of a robust library, you can often address these concerns!

React Router is a standard routing library for React applications. By using React Router, we can specify React components, that can be rendered based on the route, and so much more. Let's dive in!

### Adding A Router

> Let's make a small app to understand how this router is implemented. Create a new project using `create-react-app` and get rid of the boilerplate code as you already learned in previous lessons (just leave the `index.js` and the `App.js` files). Let's add some mock pages as an example. Create a new `Profile.js` file with the following component:

Author: Line: 79-84

2. `createBrowserRouter` is used to create the configuration for a router by simply passing arguments in the form of an array of routes.
3. The configuration array contains objects with two mandatory keys, the path and the corresponding element to be rendered.
4. This generated configuration is then rendered in, by passing it to the `RouterProvider` component.

### The Link Element

But you may notice, when we click the links in the navbar, the browser is reloading for the next URL instead of using React Router. This isn't what was promised! To help with this, [React Router exports a custom `Link` element](https://reactrouter.com/en/main/components/link) to be used instead of the regular `a` tag. We can replace the `a` tag in our navbar with the `Link` element.

~~~jsx
import { Link } from "react-router-dom";

const App = () => {
  return (
    <div>
      <h1>Hello from the main page of the app!</h1>
      <p>Here are some examples of links to other pages</p>
      <nav>
        <ul>
          <li>
            <Link to="profile">Profile page</Link>
          </li>
        </ul>
      </nav>
    </div>
  );
};

export default App;
~~~

> And now, we don't get the browser reloading everytime we click the link on the navbar!

Reviewer: ```suggestion
And now, we don't get the browser reloading every time we click the link on the navbar!
```

~~~jsx
import { Link } from "react-router-dom";

const ErrorPage = () => {
  return (
    <div>
      <h1>Oh no, this route doesn't exist!</h1>
      <Link to="/">
        You can go back to the home page by clicking here, though!
      </Link>
    </div>
  );
};

export default ErrorPage;
~~~

Add in the `errorElement` to the configuration, and check it out! The `/profile`, or any unmentioned paths render an error page!

~~~jsx
import React from "react";
import ReactDOM from "react-dom/client";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import App from "./App";
import Profile from "./Profile";
import ErrorPage from "./ErrorPage";

const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
    errorElement: <ErrorPage />,
  },
  {
    path: "profile/:name",
    element: <Profile />,
  },
]);

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>
);
~~~

### Refactoring The Routing

> But before we do that, let us refactor our routes to a component of their own, so that we can add whatever conditional logic we want, if it exists as a hook (remember, we can't use hooks outside of a React component!). Even if you don't have any need for a conditional rendering of routes, it is much neater nonetheless, to have them seperate. Create a new `Router.js` component and move your routes to it.

Reviewer: ```suggestion
But before we do that, let us refactor our routes to a component of their own, so that we can add whatever conditional logic we want, if it exists as a hook (remember, we can't use hooks outside of a React component!). Even if you don't have any need for a conditional rendering of routes, it is much neater nonetheless, to have them separate. Create a new `Router.js` component and move your routes to it.
```

import ErrorPage from "./ErrorPage";

const Router = () => {
  const router = createBrowserRouter([
    {
      path: "/",
      element: <App />,
      errorElement: <ErrorPage />,
    },
    {
      path: "profile/:name",
      element: <Profile />,
    },
  ]);

  return <RouterProvider router={router} />;
};

export default Router;
~~~

So, we can simply add this `Router.js` component to the `index.js` file.

~~~jsx
import React from "react";
import ReactDOM from "react-dom/client";
import Router from "./Router";

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <Router />
  </React.StrictMode>
);
~~~

This seems so much nicer, right?

### Protected Routes And Navigation

Often, you will face a need to decide when a certain route is rendered or not. One use case is authentication, where you may want to render certain routes based on if the user is logged in or not. If they are logged in, you may want to show some information about the user like [here at the dashboard](https://www.theodinproject.com/dashboard). While there are many ways to do so, one of the easiest ways can be to conditionally creating a config for the router.

> You will often come across the need to reroute the user to a different URL yourself. This is where the [`<Navigate />`](https://reactrouter.com/en/main/components/navigate) component gets used, as it reroutes the user to the desired URL when it is rendered. The component itself is a wrapper around [the useNavigate hook](https://reactrouter.com/en/main/hooks/use-navigate) that lets you navigate programmatically, to URLs, or even go back down the user's history.

Reviewer: "programmatically" feels a more proper word in this context:

```suggestion
You will often come across the need to reroute the user to a different URL programmatically. This is where the [`<Navigate />`](https://reactrouter.com/en/main/components/navigate) component gets used, as it reroutes the user to the desired URL when it is rendered. The component itself is a wrapper around [the useNavigate hook](https://reactrouter.com/en/main/hooks/use-navigate) that lets you navigate programmatically, to URLs, or even go back down the user's history.
```

</div>

### Knowledge Check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

- <a class="knowledge-check-link" href="#client-side-routing">What does client-side routing mean?</a>
- <a class="knowledge-check-link" href="#adding-a-router">How do you set up a basic router?</a>
- <a class="knowledge-check-link" href="#the-link-element">What should be used in place of "a" tags to enable client-side routing?</a>
- <a class="knowledge-check-link" href="#nested-routes-outlets-and-dynamic-segments">How do you create nested routes?</a>
- <a class="knowledge-check-link" href="#dynamic-segments">What do you mean by dynamic segments or URL params?</a>
- <a class="knowledge-check-link" href="#handling-bad-urls">How do you add a "catch-all" route?</a>

> - <a class="knowledge-check-link" href="#protected-routes">How do you create protected routes?</a>

Reviewer: This knowledge check link doesn't seem to work

### Additional Resources

This section contains helpful links to related content. It isn’t required, so consider it supplemental.

- Among the many ways to make protected routes, a few ways are provided below:
    - [This Stack Overflow answer](https://stackoverflow.com/a/64347082/19051112) uses a function to generate the route config object passed to `createBrowserRouter`. The function conditionally generates the different paths.

>     - [This demostration project](https://github.com/iammanishshrma/react-protected-routes/tree/master/src/components) creates a special Protected Route component that conditionally displays elements as necessary.

Reviewer: typo?
```suggestion
    - [This demonstration project](https://github.com/iammanishshrma/react-protected-routes/tree/master/src/components) creates a special Protected Route component that conditionally displays elements as necessary.
```

3. Wait for it to be nice and warm
4. Now you can eat it!

Here is where we reiterate, **the chicken needs to be reheated**. In a general multi-page application (MPAs), the browser reloads every time you click on a link to navigate. With client-side routing, **you never leave the page you are on** - you bring the microwave to the table to ensure that you don't run into the "missing spices" issues. The link requests are intercepted by the Javascript that you write, instead of letting them go directly to the server.

### A Reactive Solution

While client-side routing allows for nicer, app-like interactions (since you are controlling the routes, you can make fancy CSS animations across route changes), a lot of caveats can be missed. Browsers reloads notify screen-readers of new content to read, so you will need to notify screen-readers of route updates manually. However, with the help of a robust library, you can often address these concerns!

React Router is a standard routing library for React applications. By using React Router, we can specify React components, that can be rendered based on the route, and so much more. Let's dive in!

### Adding A Router

> Let's make a small app to understand how this router is implemented. Create a new project using `create-react-app` and get rid of the boilerplate code as you already learned in previous lessons (just leave the `index.js` and the `App.js` files). Let's add some mock pages as an example. Create a new `Profile.js` file with the following component:

Reviewer: It's decided that we will be removing CRA from the new React course. So we should remove CRA from here. Suggested wording:
```suggestion
Let's make a small app to understand how this router is implemented. Create a new React project and let's add some mock pages as an example. Create a new `Profile.js` file with the following component:
```

1. Put the chicken in the microwave and set it to cook with appropriate time and heating
2. Wait till the microwave gives out that satisfying **ding**
3. Start munching!

This is common to all websites, you set the microwave up for what you want (visit any URL, like https://theodinproject.com/), wait for the microwave to be done with the cooking (the loading screen), and tada, enjoy your delicious food (your page is ready for use). But what if you forgot to add some spices before you cooked it up? You have to repeat this flow again:

1. Add the spices to the chicken
2. Put it back into the microwave and set it up to be reheated
3. Wait for it to be nice and warm
4. Now you can eat it!

Here is where we reiterate, **the chicken needs to be reheated**. In a general multi-page application (MPAs), the browser reloads every time you click on a link to navigate. With client-side routing, **you never leave the page you are on** - you bring the microwave to the table to ensure that you don't run into the "missing spices" issues. The data is rendered with Javascript, instead of letting the browser handle it.

### A Reactive Solution

> While this allows for nicer, app-like interactions (since you are controlling the routes, you can make fancy css animations across route changes), a lot of caveats can be missed. Browsers reloads notify screen-readers of new content to read, so you will need to notify screen-readers of route updates manually. However, with the help of a robust library, you can often address these concerns!

Reviewer: Wording suggestion:
```suggestion
While client-side routing allows for nicer, app-like interactions (since you are controlling the routes, you can make fancy css animations across route changes), a lot of caveats can be missed. Browsers reloads notify screen-readers of new content to read, so you will need to notify screen-readers of route updates manually. However, with the help of a robust library, you can often address these concerns!
```

> React Router is a standard routing library for React applications. By using react-router, we can specify react components, that can be rendered based on the route.

Reviewer: ```suggestion
React Router is a standard routing library for React applications. By using react-router, we can specify react components, that can be rendered based on the route, and so much more. Let's dive in!
```

### Adding A Router

Let's make a small app to understand how this router is implemented. Create a new project using `create-react-app` and get rid of the boilerplate code as you already learned in previous lessons (just leave the `index.js` and the `App.js` files). Let's add some mock pages as an example. Create a new `Profile.js` file with the following component:

~~~jsx
const Profile = () => {
  return (
    <div>
      <h1>Hello from profile page!</h1>
      <p>So, how are you?</p>
    </div>
  );
};

export default Profile;
~~~

Replace the `App.js` file with some basic content too:

~~~jsx
const App = () => {
  return (
    <div>
      <h1>Hello from the main page of the app!</h1>
      <p>Here are some examples of links to other pages</p>
      <nav>
        <ul>
          <li>
            <a href="profile">Profile page</a>
          </li>
        </ul>
      </nav>
    </div>
  );
};

export default App;
~~~

> Now it's time to add the router! In **React Router v6.7.0**, it is recommended to add routes in form of an object. Add the following to `root.js`, we will talk about what is happening in a little bit.

Reviewer: We can future-proof ourselves a little here: 
```suggestion
Now it's time to add the router! There's a couple of ways of defining our app's routes, but in **React Router v6.7.0 or higher**, it is recommended to add routes as objects. Add the following to `root.js`, we will talk about what is happening in a little bit.
```

~~~jsx
import React from "react";
import ReactDOM from "react-dom/client";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import App from "./App";
import Profile from "./Profile";

const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
  },
  {
    path: "/profile",
    element: <Profile />,
  },
]);

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>
);
~~~

Once this is done, go ahead and run `npm start` and check out both routes: the home route "/" and the profile route "/profile" It works! But what is happening here?

1. We import `createBrowserRouter` and `RouterProvider` from "react-router-dom".
2. `createBrowserRouter` is used to create the configuration for a router by simply passing arguments in the form of an array of routes.
3. This generated configuration is then rendered in, by passing it to the `RouterProvider` component.

> 4. The configuration array contains objects with two mandatory keys, the path and the corresponding element to be rendered.

Reviewer: These points should follow our code, so I say let's swap 3 and 4.

### The Link Element

> But you may notice, when we click the links in the navbar, the browser is reloading for the next URL instead of using React Router. This isn't what was promised! To help with this, React Router exports a custom `Link` element to be used instead of the regular `a` tag. We can replace the `a` tag in our navbar with the `Link` element.

Reviewer: We could a reference link for Link:
```suggestion
But you may notice, when we click the links in the navbar, the browser is reloading for the next URL instead of using React Router. This isn't what was promised! To help with this, [React Router exports a custom `Link` element](https://reactrouter.com/en/main/components/link) to be used instead of the regular `a` tag. We can replace the `a` tag in our navbar with the `Link` element.
```

~~~jsx
import { Link } from "react-router-dom";

const App = () => {
  return (
    <div>
      <h1>Hello from the main page of the app!</h1>
      <p>Here are some examples of links to other pages</p>
      <nav>
        <ul>
          <li>
            <Link to="profile">Profile page</Link>
          </li>
        </ul>
      </nav>
    </div>
  );
};

export default App;
~~~

And now, we don't get the browser reloading everytime we click the link on the navbar!

### Nested Routes, Outlets And Dynamic Segments

Now, what if you want to render a section of a page differently, based on different URLs? This is where nested routes into play! We can add routes nested as the children of one another to ensure that the child gets rendered alongside the parent. Create a couple of components, `Popeye.js` and `Spinach.js`.

~~~jsx
import { Link } from "react-router-dom";

const Popeye = () => {
  return (
    <p>
      Hi, I am Popeye! I love to eat Spinach!{" "}
      <Link to="/">Click here to go back</Link>
    </p>
  );
};

export default Popeye;
~~~

~~~jsx
import { Link } from "react-router-dom";

const Spinach = () => {
  return (
    <p>

>       Hi, I am Spinach! Popeye loves to eat me!{" "}

Reviewer: Echoing feedback about `{" "}` from above.

      <Link to="/">Click here to go back</Link>
    </>
  );
};

export default Spinach;
~~~

Now, we can rewrite the routes as given:

~~~jsx
import React from "react";
import ReactDOM from "react-dom/client";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import App from "./App";
import Profile from "./Profile";
import Spinach from "./Spinach";
import Popeye from "./Popeye";

const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
  },
  {
    path: "profile",
    element: <Profile />,
    children: [
      { path: "spinach", element: <Spinach /> },
      { path: "popeye", element: <Popeye /> },
    ],
  },
]);

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>
);
~~~

> This allows us to render the child component alongside the parent, through an `Outlet`! We can rewrite the Profile component to add an `Outlet` which will get replaced by the various profiles when that route is visited!

Reviewer: We can add a reference link for Outlet

      { path: "popeye", element: <Popeye /> },
    ],
  },
]);

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>
);
~~~

If you visit the "/profile" path now, you should be able to see some default content where the `Outlet` is rendered when the index path is rendered!

But this example brings another dillemma. Sometimes, we want to read what was provided to the path and render content accordingly. That, here, would mean that we should be able to render content dynamically, from the component itself. Thankfully, you can do so with dynamic segments! Change the routes to be the following:

~~~jsx
import React from "react";
import ReactDOM from "react-dom/client";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import App from "./App";
import Profile from "./Profile";

const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
  },
  {
    path: "profile/:name",
    element: <Profile />,
  },
]);

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>
);
~~~

> The colon (:) has special meaning, turning the path section after it, into a <span id="dynamic-segments">"dynamic segment"</span>. Dynamic segments will match dynamic (changing) values in that position of the URL, like the `name`. These can also be called "URL params" or "params" in short. These can be used with the help of the `useParams` hook or with loaders and more, but we will use a basic example with useParams here. We can thus rewrite the profile component as the following:

Reviewer: I don't think we need to mention loaders here.

```suggestion
The colon (:) has special meaning, turning the path section after it, into a <span id="dynamic-segments">"dynamic segment"</span>. Dynamic segments will match dynamic (changing) values in that position of the URL, like the `name`. These can also be called "URL params" or "params" in short. These can be used with the help of the `useParams` hook. We can thus rewrite the profile component as the following:
```

import App from "./App";
import Profile from "./Profile";
import ErrorPage from "./ErrorPage";

const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
    errorElement: <ErrorPage />,
  },
  {
    path: "profile/:name",
    element: <Profile />,
  },
]);

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>
);
~~~

### Protected Routes

Often, you will face a need to decide when a certain route is rendered or not. One use case is authentication, where you may want to render certain routes based on if the user is logged in or not. While there are many ways to do so, one of the easiest ways can be to conditionally creating a config for the router.

But before we do that, let us refactor our routes to a component of their own, so that we can add whatever conditional logic we want, if it exists as a hook (remember, we can't use hooks outside of a React component!). Even if you don't have any need for a conditional rendering of routes, it is much neater nonetheless, to have them seperate. Create a new `Router.js` component and move your routes to it.

> 

Reviewer: Really liking the section on protected routes, but I feel like we're giving out too much code. I don't find a need to hand out implementation for protected routes for the learners here, rather letting them using external resources and figuring it out on their own. We can expand on the text explanation and add an example of a website that implements a protected route. What do you think?

Author: I don't exactly know of a website that implements this kind of protected routes, but I can look for it. Mostly, I came up with it for this new docs.

Author: From some searching, these came up as a decent couple of examples:
- https://stackoverflow.com/questions/62384395/protected-route-with-react-router-v6
  - This is mostly a way to do the protected routes and probably one of the better answers I could find. 
  - The first answer is similar to the example here, the second one and onward though, are older answers.
- https://github.com/iammanishshrma/react-protected-routes/tree/master/src/routes
  - Yet another example, this one implemented with a much granular `ProtectedRoute` component, but yes, this works.
  - This uses Redux though, which needs to be a disclaimer. The reader may see some custom hooks here.

On top of these, maybe we could refactor this to be a __Refactoring the router__ section, and a different __Protected Routes__ section?

Author: We could add another example in the form of a stackblitz thing.

Reviewer: I'm saying we should discuss protected routes as a concept. An example of a protected route would be: https://www.theodinproject.com/dashboard as it can only be accessed when a user is signed in. We don't need to give out code or any specific code implementation here, learners can figure it out themselves by researching and playing around. I hope that makes sense.

What do you have in mind for Refactoring the router section?

Author: Removed the code provided in the Protected Routes section, and added it a simple explanation, while adding a couple of implementations as an example in the additional resources section.

~~~jsx
import { useState } from "react";
import {
  createBrowserRouter,
  Navigate,
  RouterProvider,
} from "react-router-dom";
import App from "./App";
import Profile from "./Profile";
import ErrorPage from "./ErrorPage";
import Protected from "./Protected";

const Router = () => {
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  const routes = (condition) => [
    {
      path: "/",
      element: <App isLoggedIn={isLoggedIn} setIsLoggedIn={setIsLoggedIn} />,
      errorElement: <ErrorPage />,
    },
    {
      path: "profile/:name",
      element: <Profile />,
    },
    {
      path: "protected",
      element: condition ? <Protected /> : <Navigate to="/profile/popeye" />,
    },
  ];

  const router = createBrowserRouter(routes(isLoggedIn));

  return <RouterProvider router={router} />;
};

export default Router;
~~~

Here, the routes is a function that accepts a condition to see if the user is logged in. It uses that to generate a configuration of routes which are then passed on to the `createBrowserRouter` method to finally make the router.

### Assignment

<div class="lesson-content__panel" markdown="1">

> 1.  [This article](https://bholmes.dev/blog/spas-clientside-routing/) by Ben Holmes goes through a lot of the routing concepts concisely.

Reviewer: Improved link text:
```suggestion
1.  This [article on SPAs and client-side routing by Ben Holmes](https://bholmes.dev/blog/spas-clientside-routing/) goes through a lot of the routing concepts concisely.
```

2.  Go and add a few new routes to the application we created above; playing around with it is the best practice. Consider deleting it completely and rewriting it for practice.

> 3.  The [React Router tutorial](https://reactrouter.com/en/6.7.0/start/tutorial) goes through a lot of the stuff discussed in this lesson and much more. Have a read into it, and ask about, if you don't understand something. You are not expected to understand most of it in your first read.

Reviewer: This tutorial goes over a lot of extra stuff not covered in the lesson and I'm not sure if learners need to go through all of that information right now. We should narrow down what topics the learners need to cover in that tutorial.

Author: What are the specifics that should be pointed out? I think we can assume that loaders, actions and mutations, and a few more components can be pretty excessive. But I'm not sure how to tell the learner to skip these.
Guess we can explicitly mention the sections necessary and leave the learner to read the rest if they want? But there are quite a few sections, so I don't know how the can be added in a concise manner other than just listing them out.

Reviewer: Yeah, we just ask them to read from "Adding A Router" to "Nested Routes" section

![image](https://user-images.githubusercontent.com/85733202/215097822-494c5cd4-44ec-43c1-bfe0-8e9502451858.png)
